hazelcast/src/main/java/com/hazelcast/internal/networking/ChannelInboundHandler.java

```diff
@@ -16,34 +16,60 @@
+ * to be decoded into a Packet.
  *
- * {@link ChannelInboundHandler} are not expected to be thread-safe; each channel will gets its own instance(s).
+ * {@link ChannelInboundHandler} are not expected to be thread-safe; each channel
+ * will gets its own instance(s).
  *
  * A {@link ChannelInboundHandler} is constructed through a {@link ChannelInitializer}.
  *
- * If the main task of a ChannelInboundHandler is to decode a message (e.g. a Packet), it is best to call this handler a
- * decoder. For example PacketDecoder.
+ * If the main task of a ChannelInboundHandler is to decode a message (e.g. a Packet),
+ * it is best to call this handler a decoder. For example PacketDecoder.
  *
  * @see ChannelOutboundHandler
  * @see EventLoopGroup
  * @see ChannelInitializer
  * @see ChannelErrorHandler
  * @see Channel
  */
-public interface ChannelInboundHandler {
+public abstract class ChannelInboundHandler<S, D> extends ChannelHandler<ChannelInboundHandler, S, D> {
 
     /**
-     * A callback to indicate that data is available in the src ByteBuffer to be processed.
+     * A callback to indicate that data is available in the src ByteBuffer to be
+     * processed.
      *
-     * @param src the ByteBuffer containing the data to read. The ByteBuffer is already in reading mode and when completed,
-     *            should not be converted to write-mode using clear/compact.
-     * @throws Exception if something fails while reading data from the ByteBuffer or processing the data (e.g. when a Packet
-     *                   fails to get processed). When an exception is thrown, the {@link ChannelErrorHandler} is called.
+     * ChannelInboundHandlers should be able to deal with spurious onReads.
+     *
+     * The ByteBuffer containing the data to read. The ByteBuffer is already in
+     * reading mode and when completed, should not be converted to write-mode using
+     * clear/compact.
+     *
+     * @return true when read everything that can be read is read. False if data
+     * remaining and the rest of the pipeline needs to be processed first before
+     * retrying the pipeline.
+     * @throws Exception if something fails while reading data from the ByteBuffer
+     * or processing the data (e.g. when a Packet fails to get processed). When an
+     * exception is thrown, the {@link ChannelErrorHandler} is called.
+     */
+    public abstract HandlerStatus onRead() throws Exception;
+
+    /**
+     * Initializes the src buffer. Should only be called by ChannelInboundHandler
+     * implementations that have a ByteBuffer as source.
      */
-    void onRead(ByteBuffer src) throws Exception;
+    protected final void initSrcBuffer() {
+        initSrcBuffer(channel.config().getOption(SO_RCVBUF));
+    }
+
+    protected final void initSrcBuffer(int sizeBytes) {
+        ChannelConfig config = channel.config();
+        src = (S) newByteBuffer(sizeBytes, config.getOption(DIRECT_BUF));
```

Comment:
If source is always `ByteBuffer`, why `S` is used/needed?

Reply:
Src doesn't need to be a ByteBuffer. This method should only be made if the ChannelinboundHandler src should be a bytebuffer.